words/headers.asm

```diff
@@ -8,33 +8,40 @@
 ; tricks in the code. We roughly follow the Gforth terminology: The Execution
 ; Token (xt) is the address of the first byte of a word's code that can be, uh,
 ; executed; the Name Token (nt) is a pointer to the beginning of the word's
-; header in the Dictionary. There the link to the next word in the Dictionary
-; is always one cell down from the current word's own nt. In the code itself,
-; we use "nt_<WORD>" for the nt and "xt_<WORD>" for the xt.
-
-; This gives us the following header structure:
-
-;              8 bit     8 bit
-;               LSB       MSB
-; nt_word ->  +--------+--------+
-;          +0 | Length | Status |
-;             +--------+--------+
-;          +2 | Next Header     | -> nt_next_word
-;             +-----------------+
-;          +4 | Start of Code   | -> xt_word
-;             +-----------------+
-;          +6 | End of Code     | -> z_word
-;             +--------+--------+
-;          +8 | Name   |        |
-;             +--------+--------+
-;             |        |        |
-;             +--------+--------+
-;             |        |  ...   | (name string does not end with a zero)
-;          +n +--------+--------+
-
-; The Status Byte is created by adding the flags defined in definitions.asm,
-; which are:
-
+; header in the Dictionary.   Conceptually the header looks like this:
+
+;   nt_word --> +----------------+
+;               | Status flags   |
+;               +----------------+
+;               | Name length    |
+;               +----------------+
+;               | Last header    | -> nt_last_word
```

Comment:
My brain has a real issue with the word "last" here - all I can think about is "final".  Would "Previous header" and "nt_prev_word" work here?  I think that also carries the appropriate connotation that the previous word is often earlier in memory.

Reply:
100% agree.  i struggled a lot with how to describe the ordering of the words since you have competing concepts of order in memory, order of the linked list and order of the dictionary.   The last two always seem opposite to me since the first word you check in the dictionary is the last word in a backward-linked list.

I'm happy to switch to previous for the header link since that matches my intuition for the linked list.  The words higher / lower might work for memory ordering.   Is there a good way to talk about dictionary order that doesn't result in complete confusion with linked list terms?

Reply:
I think just being clear in both the comments and documentation is the best we can do.  I'll have to review the docs, but I believe it has a reasonably clear explanation that words are searched from newest to oldest.  I think we should avoid using "start" and "end" to talk about the list, perhaps using "latest" to talk about the most recent word and "the oldest word" to talk about the word where previous=NULL.  Many of the loops accept null as the nt (an empty wordlist, for example) and keep looping until the current nt is null - in that case we could simply say "until there are no more words".

I think just being mindful that this is an issue and choosing our words carefully should be sufficient.  All Forths generally have this backward list notion as they are required to find the newest version of a word if it is redefined.

"Previous" is nice here because it implies both "already existing before this word" as well as "earlier in memory".

Also, just FYI, the headers used to be the other way around but they were rearranged to allow for easier optional words with conditional compilation.